skimage/morphology/_flood_fill.py

```diff
@@ -0,0 +1,290 @@
+    >>> image
+    array([[5, 5, 5, 5, 5, 5, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 5, 5, 3]])
+    """
+    # Correct start point in ravelled image - only copy if non-contiguous
+    image = np.asarray(image)
+    if image.flags.f_contiguous is True:
+        order = 'F'
+    elif image.flags.c_contiguous is True:
+        order = 'C'
+    else:
+        warnings.warn('Non-contiguous array passed as `image`; this will be '
+                      'converted to a contiguous array as a copy.')
+        image = np.ascontiguousarray(image)
+        order = 'C'
+
+    seed_value = image[seed_point]
+
+    # Shortcut for rank zero
+    if image.size == 0:
+        return np.array([], dtype=np.uint8)
+
+    # Convenience for 1d input
+    try:
+        iter(seed_point)
+    except TypeError:
+        seed_point = (seed_point,)
+
+    selem = _resolve_neighborhood(selem, connectivity, image.ndim)
+
+    # Must annotate borders
+    working_image = _fast_pad(image, image.min())
+
+    # Stride-aware neighbors - works for both C- and Fortran-contiguity
+    ravelled_seed_idx = np.ravel_multi_index([i+1 for i in seed_point],
+                                             working_image.shape, order=order)
+    neighbor_offsets = _offsets_to_raveled_neighbors(
+        working_image.shape, selem, center=((1,) * image.ndim))
+
+    # Use a set of flags; see _flood_fill_cy.pyx for meanings
+    flags = np.zeros(working_image.shape, dtype=np.uint8)
+    _set_edge_values_inplace(flags, value=2)
+
+    try:
+        if tolerance is not None:
+            # Check if tolerance could create overflow problems
+            try:
+                max_value = np.finfo(working_image.dtype).max
+                min_value = np.finfo(working_image.dtype).min
+            except ValueError:
+                max_value = np.iinfo(working_image.dtype).max
+                min_value = np.iinfo(working_image.dtype).min
+
+            high_tol = min(max_value, seed_value + tolerance)
+            low_tol = max(min_value, seed_value - tolerance)
+
+            _flood_fill_tolerance(working_image.ravel(),
+                                  flags.ravel(),
+                                  neighbor_offsets,
+                                  ravelled_seed_idx,
+                                  seed_value,
+                                  low_tol,
+                                  high_tol)
+        else:
+            _flood_fill_equal(working_image.ravel(),
+                              flags.ravel(),
+                              neighbor_offsets,
+                              ravelled_seed_idx,
+                              seed_value)
+    except TypeError:
+        if working_image.dtype == np.float16:
+            # Provide the user with clearer error message
+            raise TypeError("dtype of `image` is float16 which is not "
+                            "supported, try upcasting to float32")
```

Comment:
Why is this the case?

Reply:
Float16 is not among the fused types in Cython, so catch and pass an informative error.  Personally I feel float16 is dangerously low precision warranting its exclusion; precision error propagates fast.  We excluded float16 in the initial code supported by @lagru also.